### ISSUE TITLE
composer ignores psr-4 path

### ISSUE THREAD

**Issue by dspiegel** [composer/composer#4132]
Trying to update from psr-0 to psr-4, but not working like its supposed to. Followed instructions on composer site, but now my bundle is installed without designated path. What is wrong?

##### psr-0:
```javascript
     "name": "dst/app-bundle",
     "target-dir":  "Dst/AppBundle",
     "autoload":    {
          "psr-0": {
             "Dst\\AppBundle\\": ""
           }
      },
```
   **Bundle gets installed in:** *vendor/dst/app-bundle/Dst/Appbundle*

##### psr-4:
```javascript
     "name": "dst/app-bundle",
     "autoload":    {
          "psr-4": {
             "Dst\\AppBundle\\": ""
           }
      },
```
   **Bundle gets installed in:** *vendor/dst/app-bundle*

Composer version 1.0-dev (4f80e7ff68466cab970c22b425725b8058c32970) 2015-06-09 00:03:48

**Comment by nevvermind** [comment]
I'm not an expert in PSR4, but that looks fine to me. Create `vendor/dst/app-bundle/MyClass.php` and use it like `new Dst\AppBundle\MyClass()`. Does it work? 'Cause it should.

> but not working like its supposed to

A little vague there.

**Comment by dspiegel** [comment]
@nevermind I guess I should say it does not work like it did previously or like any other bundle that are installed such as Symfony. 

Previously the bundle was installed in **vendor/dst/app-bundle/Dst/AppBundle** folder, but now using same setting it installs in **vendor/dst/app-bundle** folder. The autoload_psr4.php  then fails because it cannot find Dst/AppBundle/DstAppBundle.php.

 autoload_psr4.php has the following:

```javascript
    'Dst\\AppBundle\\' => array($vendorDir . '/dst/app-bundle/Dst/AppBundle'),
```

Because the bundle was **installed in vendor/dst/app-bundle not in vendor/dst/app-bundle/Dst/AppBundle** it cannot be autoloaded. I just want it to work like it did before and as the composer psr-0 to psr-4 update example indicates it should.

I checked other bundles that are loaded and they are defined just like mine, except with "src/" as path. I tried that too but still drops files in namespace (dst/app-bundle) folder.

**Comment by nevvermind** [comment]
Have you ran `composer dumpautoload` after changing to PSR4? Try `rm -rf vendor/dst && composer require dst/app-bundle` again. `vendor/dst/app-bundle/Dst/AppBundle` doesn't look right.

**Comment by slbmeh** [comment]
A PSR-4 install shouldn't have a target-dir.  The target-dir makes it possible to have a PSR-4 bundle structure with a PSR-0 autoloader.

https://github.com/symfony/FrameworkBundle/commit/e18c83dca02469c08432bd4cd2bea278f2b688c9

@dspiegel What is your desired directory structure, and how does that compare to your namespace?

**Comment by alcohol** [comment]
I think you should report this issue at [composer/installers](https://github.com/composer/installers).

**Comment by stof** [comment]
I think the issue is that you don't understand what the PSR-4 spec defines. The structure you describe is eprfectly fine for PSR-4.

**Comment by dspiegel** [comment]
Thanks for all the feedback. I've tried everything that has been suggested, but still no solution.

@slbmeh I did remove target-dir for PSR-4. The desired directory structure would be the same one that PSR-0 with tagert-dir created:
`vendor/dst/app-bundle/Dst/AppBundle`

@stof I agree I definitely don't understand PSR-4. Why my bundle installs differently than all other Bundles which use same PSR-4 structure I am using.

For instance using Symfony's autload definition as example:
```javascript
 "name": "symfony/symfony",
 "autoload": {
        "psr-4": {
             "Symfony\\Bundle\\": "src/Symfony/Bundle/"
  },
```
  Installs Symfony bundle in **vendor/symfony/symfony/src/Symfony/Bundle**

If I do the same for my bundle:

```javascript
  "name":        "dst/app-bundle",
  "autoload":    {
       "psr-4": {
            "Dst\\AppBundle\\":  "src/DstAppBundle/"                    
        }
   },

```
It gets installed in **vendor/dst/app-bundle**

So hopefully this will show why I am confused. Also as I indicated before, the composer/autoload_ps4.php would define above as 
```javascript
'Dst\\AppBundle\\' => array($vendorDir . '/dst/app-bundle/src/Dst/AppBundle'),
```
just as I would expect it to, but my Bundle did not get installed in 'src'/Dst/AppBundle' so *Failed to find class* error occurs.

I'm not sure how I can make this any clearer as to why I feel it's not working properly or maybe I have missed something in config that makes Symfony bundle install differently than my bundle.

Here is my root composer.json:
```javascript
{
    "name":              "dst/external-iprd",
    "description":       "Dst Application",
    "license":           "proprietary",
    "require":           {
        "dst/app-bundle": "*"
    },
    "require-dev":       {
        "sensio/generator-bundle": "~2.3"
    },
    "autoload":          {
        "psr-4": {
            "": "src/"
        }
    },
    "scripts":           {
        "post-root-package-install": [
            "SymfonyStandard\\Composer::hookRootPackageInstall"
        ],
        "post-install-cmd":          [
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::buildBootstrap",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::clearCache",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installAssets",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installRequirementsFile",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::removeSymfonyStandardFiles",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::prepareDeploymentTarget"
        ],
        "post-update-cmd":           [
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::buildBootstrap",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::clearCache",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installAssets",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::installRequirementsFile",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::removeSymfonyStandardFiles",
            "Sensio\\Bundle\\DistributionBundle\\Composer\\ScriptHandler::prepareDeploymentTarget"
        ]
    },
    "config":            {
        "bin-dir":           "bin",
        "preferred-install": "dist"
    },
    "minimum-stability": "stable",
    "extra":             {
        "symfony-app-dir":        "app",
        "symfony-web-dir":        "web",
        "symfony-assets-install": "relative"
    },
    "repositories":      [
        {
            "type":     "composer",
            "url": "http://svnserver/svn/vendor/Symfony2/DstBundleManager/trunk
        }
    ]
}

```

Here is my current packages.json file (not using "src" folder as shown in above example)
```javascript
{
    "packages": {
        "dst/app-bundle":  {
            "production": {
                "name":        "dst/app-bundle",
                "type":        "symfony-bundle",
                "version":     "2015.06.10.1633",
                "description": "DstAppBundle, DST Group application bundle",
                "license":     "proprietary",
                "source":      {
                    "type":      "svn",
                    "url":       "http://svnserver/svn/vendor/Symfony2/DstAppBundle",
                    "reference": "/trunk/@16325"
                },
                "autoload":    {
                    "psr-4": {
                        "Dst\\AppBundle\\":  "",
                        "SymfonyStandard\\": "app/"
                    }
                },
                "require":     {
                    "php":                           ">=5.5.9",
                    "symfony/symfony":               "2.7.*",
                    "doctrine/orm":                  "~2.2,>=2.2.3,<2.5",
                    "doctrine/dbal":                 "<2.5",
                    "doctrine/doctrine-bundle":      "~1.4",
                    "symfony/assetic-bundle":        "~2.3",
                    "symfony/swiftmailer-bundle":    "~2.3",
                    "symfony/monolog-bundle":        "~2.4",
                    "sensio/distribution-bundle":    "~3.0,>=3.0.12",
                    "sensio/framework-extra-bundle": "~3.0,>=3.0.2",
                    "jms/security-extra-bundle":     "1.5.*",
                    "knplabs/knp-menu":              "2.0.1",
                    "knplabs/knp-menu-bundle":       "2.0.0",
                    "phing/phing":                   "~2.11"
                },
                "require-dev": {
                    "sensio/generator-bundle": "~2.3"
                },
                "bin":         [
                    "bin/init",
                    "bin/phpDocumentor.phar",
                    "bin/yuicompressor.jar"
                ]
            }
        }
    }
}
```

**Comment by jvasseur** [comment]
The strange thing here is not were your bundle is installed, that's the correct behaviour but the content of your autoload_psr4.php that should be ``` 'Dst\\AppBundle\\' => array($vendorDir . '/dst/app-bundle/src')```

**Comment by dspiegel** [comment]
I'm going to close this issue as I believe I have some structure problem with my Bundles I need to resolve. I believe my initial problem was created using the PSR-0 target-dir which allowed bundle to work even though I built a bundle dir structure that does not match the Namespace.

Thanks for all the feedback, it was very helpful.

- Dave

**Comment by stof** [comment]
@dspiegel if you don't have a ``src/Dst/AppBundle/`` folder **inside** your package, you should not use such prefix in your composer autoloading config. If you have the bundle class at the root of the repo (setup for PSR-0 + target-dir), the config should be 

```json
{
  "name":        "dst/app-bundle",
  "autoload":    {
       "psr-4": {
            "Dst\\AppBundle\\":  ""                    
        }
   }
}
```

But anyway, without seeing the folder structure of your repo, it is quite impossible to help you to get the right config

**Comment by alcohol** [comment]
If you are able to share with us your resulting directory structure from the issue above;

```
tree -d -x -l /path/to/project
```

**Comment by dspiegel** [comment]
With everyone's help I now understand what went wrong.

1. Several years ago, I created a bundle under a projects src folder (src/Dst/AppBundle).
2. I moved the stable bundle to standalone definition in it's own local repository using packages.json which pulls code from SVN. 
  - Now it gets deployed using composer and is used in all my teams projects without any additional setup on their part.
3. When I separated the bundle in SVN, I did not keep 'Dst/AppBundle' folder as part of SVN dir structure. It was just the root bundle.
4. I added "target-dir" = "Dst/AppBundle" to packages.json definition to force install into proper folder to allow Class namespace to be correct. This was probably incorrect decision on my part.
5. PSR-4 does not support "target-dir", so it placed my Bundle in the correct location "/vendor/dst/app-budle" because I do not have dir structure in my project that matches the Class namespace.

I am going to restructure the Bundle in SVN using 'src/Dst/AppBundle', but continue using PSR-0 for now.

My main issue with changing to PSR-4 is that bundle adds several links to bin folder to support application. Since composer will not remove old links and replace with new link location, I will have to have each team member remove their links manually before upgrading to Symfony 2.7. My hope was to do upgrade with minimal or no adjustments by programmer.

Hope this explains my problem completely. If any one has suggestions on a better way to resolve issue, I would be happy to hear them.

Thanks, Dave

**Comment by slbmeh** [comment]
@dspiegel I'm not I fully understand the issue you're having.

PSR-4 should make it so you don't need a corresponding directory structure and allows you to do exactly what you did without the `target-dir` parameter to make it work.

If you have `Dst\AppBundle\DstAppBundle` in the root folder with `"Dst\\AppBundle\\": ""` for your PSR-4 autoload, it should be able to find the class located in `vendor/dst/app-bundle/DstAppBundle.php`

**Comment by dspiegel** [comment]
@slbmeh That's the exact problem that I noted which I found, I don't have `Dst\AppBundle\DstAppBundle` in my root folder, just have the DstAppBundle.php file and it's sub folders. That is change I need to do to my bundle to make it work properly with PSR-4..

**Comment by jvasseur** [comment]
@dspiegel This looks wrong to me, you don't need the ```Dst/AppBundle/``` directory structure into your repository to work with psr-4.

Look at the MonologBundle (https://github.com/symfony/MonologBundle) for an example.

**Comment by slbmeh** [comment]
@dspiegel you don't need the directories... that's the point of PSR-4

When you declare PSR-0 `"Dst\\AppBundle\\": "src/"`, that means that every class starting with that namespace is found under `src/`, and in PSR-0 each segment of the namespace is represented by a directory.

Whereas, with PSR-4, that means the same as far as the location goes, but the namespace segments that represent directories begin after `Dst\\AppBundle\\`, which having `DstAppBundle.php` top level is exactly where it should find `Dst\\AppBundle\\DstAppBundle` if you are using PSR-4.

Basically if you don't define a namespace for your autoloads PSR-0 and PSR-4 should behave and appear the same, when you do declare a namespace PSR-0 is expecting to find those folders and PSR-4 will ignore the namespace prefix and start looking right at the path you defined.
